crates/router/src/db/api_keys.rs

```diff
@@ -126,55 +126,254 @@ impl ApiKeyInterface for Store {
-        // [#172]: Implement function for `MockDb`
-        Err(errors::StorageError::MockDbError)?
+        Ok(self
+            .api_keys
+            .lock()
+            .await
+            .iter()
+            .find(|k| k.merchant_id == merchant_id && k.key_id == key_id)
+            .cloned())
     }
 
     async fn find_api_key_by_hash_optional(
         &self,
-        _hashed_api_key: storage::HashedApiKey,
+        hashed_api_key: storage::HashedApiKey,
     ) -> CustomResult<Option<storage::ApiKey>, errors::StorageError> {
-        // [#172]: Implement function for `MockDb`
-        Err(errors::StorageError::MockDbError)?
+        Ok(self
+            .api_keys
+            .lock()
+            .await
+            .iter()
+            .find(|k| k.hashed_api_key == hashed_api_key)
+            .cloned())
     }
 
+    #[allow(clippy::unwrap_used)]
     async fn list_api_keys_by_merchant_id(
         &self,
-        _merchant_id: &str,
-        _limit: Option<i64>,
-        _offset: Option<i64>,
+        merchant_id: &str,
+        limit: Option<i64>,
+        offset: Option<i64>,
     ) -> CustomResult<Vec<storage::ApiKey>, errors::StorageError> {
-        // [#172]: Implement function for `MockDb`
-        Err(errors::StorageError::MockDbError)?
+        // first gather all the keys for the given merchant_id, then apply any limit or offset, if
+        // requested. This uses memory inefficiently if a limit/offset is requested, but should be
+        // sufficient for the mockdb usecase
+        let mut keys_for_merchant_id: Vec<storage::ApiKey> = self
+            .api_keys
+            .lock()
+            .await
+            .iter()
+            .filter(|k| k.merchant_id == merchant_id)
+            .cloned()
+            .collect();
+
+        // mimic the SQL limit/offset behavior
+        if let Some(offset) = offset {
+            if offset < 0 {
+                Err(errors::StorageError::MockDbError)?;
+            }
+            let offset: usize = offset.try_into().unwrap();
+            if offset >= keys_for_merchant_id.len() {
+                keys_for_merchant_id = vec![];
+            } else {
+                keys_for_merchant_id = keys_for_merchant_id[offset..].to_vec();
+            }
+        }
+
+        if let Some(limit) = limit {
+            if limit < 0 {
+                Err(errors::StorageError::MockDbError)?;
+            }
+            let limit: usize = limit.try_into().unwrap();
+            keys_for_merchant_id.truncate(limit);
+        }
+
+        Ok(keys_for_merchant_id)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use time::macros::datetime;
+
+    use crate::{
+        db::{api_keys::ApiKeyInterface, MockDb},
+        types::storage,
+    };
+
+    // ignored for now because we need to do some finagling with redis to get it to run, since the
+    // mockdb needs to connect to a real redis instance
+    #[tokio::test]
+    #[ignore]
```

Comment:
Could you remove the `#[ignore]` annotation now that you've opened the PR?

Reply:
done